design/subscriber-sync.feature

```diff
@@ -0,0 +1,64 @@
+Feature: Newsletter Subscriber synchronization
+  - As a store owner
+  - I want to update subscribers information automatically
+  - So that I can save time on my newsletter preparations
+
+  Entities:
+    Admin - A person that has admin rights in Magento store.
+    Visitor - A person that visits Magento store but is anonymous.
+    Guest - A person who has created an order or signed up for a newsletter. Guest has not created
+      unique customer account, but has a distinct email address in Magento store.
+    Customer - A person who has created an account in store.
+    Client - A person who is a guest or a customer.
+    Subscriber - Person who has at one point joined newsletter list.
+    Unsubscriber - Person who has canceled their subscription.
```

Comment:
So Magneto has established a few groups by itself. Admin, Guest, and Customer - so I didn't want to change those.
Created new groups:
Visitor - anonymous person. Looks around but never buys anything
Client - a person who exists in Magento but in our context there is no need to differentiate between customer and guest. I once bought a pair of sandals dude. It comes into play when subscriber status is not set and there is a status change happening.

Reply:
I think it shouldn't matter if Magento already has customer groups with the terms we would like to use. The Entities section should define the terms the design document uses. If the term entails multiple Magento customer groups then it can state that in its definition.

An issue that I see with some terms is that they are compounds of different terms (for example Client). I think we should try to avoid using compound terms and solve it by using different scenarios for each entity.

Reply:
One more thing regarding Magento's customer groups - I think administrator can always change them and add more, if needed. In these cases relying on Magento's customer groups might not work.

Reply:
> One more thing regarding Magento's customer groups - I think administrator can always change them and add more if needed. In these cases relying on Magento's customer groups might not work.

So there are two types of "subscribers" that we need to sync and these are defined not by customer groups but as is. Guest and customer. Customers can be set to different named groups but this does not concern us in the plugin context.
![image](https://user-images.githubusercontent.com/22918359/70533810-7f04fe00-1b62-11ea-8179-80f4fdeab20a.png)

> I think it shouldn't matter if Magento already has customer groups with the terms we would like to use. The Entities section should define the terms the design document uses. If the term entails multiple Magento customer groups then it can state that in its definition.
> 
> An issue that I see with some terms is that they are compounds of different terms (for example Client). I think we should try to avoid using compound terms and solve it by using different scenarios for each entity.

The issue with that approach may arise when we want to use the same terms in the context of other features. For example, when using the term `Customers`  for both `Guest` and `Customer`. When we diverge from the terms Magento already provides and use the same term as our own we may run into a clash that we have to create a term for something that is defined by Magento
>  Scenario: Guest subscribes
    Given guest fills newsletter sign-up form
    When guest submits the form
    Then welcome email is sent

And 
>   Scenario: Customer subscribes
    Given customer fills newsletter sign-up form
    When customer submits the form
    Then welcome email is sent

This would actually work as there is no different outcome.

Reply:
So I looked into most of the cases. The guest and customer group could be used as a customer. Most of the results are the same for both in our context. The only stupid logic is Magento-s own when the validation for signup separates both. 

  >Scenario: Guest subscribes with an email address belonging to newsletter subscriber
    Given guest fills newsletter sign-up form
    And email address has already been subscribed to newsletter
    When guest submits the form
    Then no welcome email is sent

  >Scenario: Guest subscribes with an email address belonging to customer
    Given guest fills newsletter sign-up form
    And email address already belongs to a customer
    When guest submits the form
    Then error message is shown

This could also be resolved as the result is actually the same, one receives extra error.

Reply:
I changed a lot again in  https://github.com/sendsmaily/Sendsmaily-Sync-for-Magento/pull/8/commits/bd6764e841722709268a05fb385ce59dcf8f0c36 
I removed most of the Entities and I think it is the simplest and most descriptive solution I can come up with about how subscriptions should be updated.

I would also keep the subscription logic here. ( Update subscribers state only with this feature)
And add subscribers data gathering as a separate feature
>As a store owner
I would like to get subscribers personal information
So that I would be able to send personalized emails

Or something like that